File: app/src/main/java/cn/nicolite/huthelper/network/exception/ExceptionEngine.java
package cn.nicolite.huthelper.network.exception;


import com.google.gson.JsonParseException;
import com.google.gson.stream.MalformedJsonException;

import org.json.JSONException;

import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.text.ParseException;

import retrofit2.HttpException;

/**
 * 错误/异常处理工具
 * Created by nicolite on 17-10-13.
 */

public class ExceptionEngine {
    public static final int UN_KNOWN_ERROR = 1000;//未知错误
    public static final int PARSE_SERVER_DATA_ERROR = 1001;//解析(服务器)数据错误
    public static final int PARSE_CLIENT_DATA_ERROR = 1002;//解析(客户端)数据错误
    public static final int CONNECT_ERROR = 1003;//网络连接错误
    public static final int CONNECT_TIME_OUT_ERROR = 1004;//网络连接超时

    public static APIException handleException(Throwable throwable){
        throwable.printStackTrace();

        APIException apiException;
        if (throwable instanceof HttpException){
            HttpException httpException = (HttpException) throwable;
            apiException = new APIException(throwable, httpException.code());
            apiException.setMsg("网络错误");
            return apiException;
        }else if (throwable instanceof JsonParseException
                || throwable instanceof JSONException
                || throwable instanceof ParseException
                || throwable instanceof MalformedJsonException){
            apiException = new APIException(throwable, PARSE_SERVER_DATA_ERROR);
            apiException.setMsg("解析数据错误");
            return apiException;
        }else if (throwable instanceof ConnectException){
            apiException = new APIException(throwable, CONNECT_ERROR);
            apiException.setMsg("网络连接失败");
            return apiException;
        }else if (throwable instanceof SocketTimeoutException){
            apiException = new APIException(throwable, CONNECT_TIME_OUT_ERROR);
            apiException.setMsg("网络连接超时");
            return apiException;
        }else if (throwable instanceof ServerException){
            ServerException serverException = (ServerException) throwable;
            apiException = new APIException(serverException, serverException.getCode());
            apiException.setMsg(serverException.getMsg());
            return apiException;
        } else {
            apiException = new APIException(throwable, UN_KNOWN_ERROR);
            apiException.setMsg("未知错误");
            return apiException;
        }
    }
}

File: app/src/main/java/cn/nicolite/huthelper/exception/ExceptionEngine.kt
package cn.nicolite.huthelper.exception


import com.google.gson.JsonParseException
import com.google.gson.stream.MalformedJsonException
import org.json.JSONException
import retrofit2.HttpException
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.text.ParseException

/**
 * 错误/异常处理工具
 * Created by nicolite on 17-10-13.
 */

enum class ExceptionError(val code: Int, val msg: String) {
    UN_KNOWN_ERROR(100, "未知错误"),
    PARSE_SERVER_DATA_ERROR(101, "解析(服务器)数据错误"),
    PARSE_CLIENT_DATA_ERROR(102, "解析(客户端)数据错误"),
    CONNECT_ERROR(103, "网络连接错误"),
    CONNECT_TIME_OUT_ERROR(104, "网络连接超时"),
    SERVER_ERROR(105, "服务器错误"),
    HTTP_ERROR(106, "网络错误")
    ;
}

fun handleAPIException(throwable: Throwable): APIException {
    throwable.printStackTrace()
    val apiException: APIException
    when (throwable) {
        is HttpException -> {
            apiException = APIException(ExceptionError.HTTP_ERROR)
        }
        is retrofit2.adapter.rxjava2.HttpException -> {
            apiException = APIException(ExceptionError.HTTP_ERROR)
        }
        is JsonParseException -> {
            apiException = APIException(ExceptionError.PARSE_SERVER_DATA_ERROR)
        }
        is JSONException -> {
            apiException = APIException(ExceptionError.PARSE_SERVER_DATA_ERROR)
        }
        is ParseException -> {
            apiException = APIException(ExceptionError.PARSE_SERVER_DATA_ERROR)
        }
        is MalformedJsonException -> {
            apiException = APIException(ExceptionError.PARSE_SERVER_DATA_ERROR)
        }
        is ConnectException -> {
            apiException = APIException(ExceptionError.CONNECT_ERROR)
        }
        is SocketTimeoutException -> {
            apiException = APIException(ExceptionError.CONNECT_TIME_OUT_ERROR)
        }
        else -> {
            apiException = APIException(ExceptionError.UN_KNOWN_ERROR)
        }
    }
    return apiException
}

